src/components/CharacterAssignment.jsx

```diff
@@ -0,0 +1,229 @@
+import UnassignedColumn from './UnassignedColumn';
+
+// Entire Drag and Drop context for assigning characters to actors
+const CharacterAssignment = ({actors, currentScript}) => {
+  const [columnOrder, setColumnOrder] = useState([])
+  const [columns, setColumns] = useState({
+    'unassignedCharacters': {
+      id:'unassignedCharacters',
+        title: 'Unassigned Characters',
+        characterIds: [],
+        lineCount: 0
+    }});
+  const [characters, setCharacters] = useState()
+
+  // Create the actor columns initially and edit columns when actors are added or removed
+  useEffect(() => {createOrUpdateActorColumns(actors)}, [actors]);
+
+  // get the character data on the initial load and when the script option changes
+  useEffect(() => { (async () => { // useEffect cannot take an async function. Must wrap async in regular function
+    if (currentScript === null) {
+      return;
+    }
+    try {
+      const response = await fetch('/api/script/' + currentScript + '/characters');
+
+      if (response.ok) {
+        const fetchedCharacters = await response.json()
+        resetActors()
+        setCharacters(fetchedCharacters);
+        setAssignedCharacters(fetchedCharacters);
+
+      } else {
+        console.error(`server error: ${response.body} when fetching character data`);
+      }
+    } catch (error) {
+      console.error(`network error: ${error} when fetching character data`);
+    }
+  })()}, [currentScript]);
+
+
+
+  // Reset the assigned characters to the actors so we don't have old character Ids interfering
+  const resetActors = () => {
+    const newColumns = {...columns}
+    for (let key in newColumns) {
+      newColumns[key].characterIds = [];
+      newColumns[key].lineCount = 0;
+    }
+    setColumns({...newColumns})
+  }
+
+
+  // Set all characters in their assigned column based on the fetched data
+  const setAssignedCharacters = (fetchedCharacters) => {
+    const newColumns = {...columns}
+
+    for (let key in fetchedCharacters){
+      let assignedActor = fetchedCharacters[key].actorId
+      if (assignedActor) {
+        newColumns[assignedActor].characterIds.push(fetchedCharacters[key].id)
+        newColumns[assignedActor].lineCount += fetchedCharacters[key].lineCount
+      } else {
+        newColumns.unassignedCharacters.characterIds.push(fetchedCharacters[key].id)
+        newColumns.unassignedCharacters.lineCount += fetchedCharacters[key].lineCount
+      }
+    }
+    setColumns({...newColumns})
+  }
+
+  // Create and update the columns for each of the actors and an Unassigned Characters column
+  const createOrUpdateActorColumns = (actors) => {
+    // Script is already set, only add or remove actors
+    if (currentScript && actors.length > 1  ) {
+      const newColumns = {...columns}
+      const newOrder = [...columnOrder]
+      // Adding an actor, there are more actors than columns (plus the unassigned column)
+      if (actors.length + 1 > newOrder.length){
```

Comment:
might consider extracting the bodies of these `if/else`s into their own functions so the control flow is easier to see